sql_grader/problem.py

```diff
@@ -36,20 +38,28 @@ def __init__(
         self.is_ordered = is_ordered
         self.answer_query = answer_query
         self.verify_query = verify_query
-        self.answer_result, _ = SqlProblem.run_query(
+        self.modification_query = modification_query
+        self.answer_result, self.answer_error = SqlProblem.run_query(
             self.database,
             answer_query,
             verify_query,
+            modification_query,
         )
 
     def attempt(self, query):
         """
         Attempt to answer the problem with the provided query
         """
+        if self.answer_error:
```

Comment:
True, there's a difference between detecting SQL errors in Studio vs. when running queries.
I don't know how it would be done from Studio: whether there's enough information to actually run the SQL (the submitted answer will be missing at this point). And whether the slowness that we see when running the query in the LMS could cause some effect in Studio. And whether Studio's error reporting workflow is convenient.
I'd need more time to take a look in detail. Unfortunately I'll be off for the next 2 weeks so it would happen later. I'll take a fast look then to see whether this change is easy.

Reply:
@stvstnfrd I took a closer look.

- Currently, Studio is used to set up the block, but it just saves the queries as configuration. It doesn't run them in sqlite. This is because the `answer_query` is first run [in the SqlProblem constructor](https://github.com/open-craft/xblock-sql-grader/blob/56b13ea9719d5c6e910ec2edf98b35dda0ba9c57/sql_grader/problem.py#L39), and SqlProblem is only instantiated in one place: when [grading](https://github.com/open-craft/xblock-sql-grader/blob/56b13ea9719d5c6e910ec2edf98b35dda0ba9c57/sql_grader/mixins/grading.py#L40)
- If staff members make a typo when writing answer_query/verify_query/modification_query in Studio, the queries are saved. When trying to submit an answer, students will see "Problem setup incorrectly".
- We could change it to this behaviour: every time the block is edited, the dataset is loaded, the queries (answer_query/verify_query/modification_query) are run and validated. If there are SQL errors, the block can't be save. Note that the student's query isn't involved or needed because it doesn't exist at this point
- I find 2 issues with the proposal:
  - Speed: building the sqlite DB, the code jail, and running the queries takes a long time (>30 seconds in our sandbox). This would happen every time the block is edited in Studio. I find it would be easier if: whenever staff wants to test the block, they test it as a student. This shouldn't be needed often, just after editing queries.
  - UI: we'd need to block the save event in Studio and display errors detailing which fields are wrong and why, and prevent the save. We'd also need to add a spinner after the Save button is clicked, while the query is being run.
- I didn't check Studio's UI capabilities, but the speed reduction alone doesn't make it very attractive. What do you think? In any case, the UI changes and new testing could mean >10 h more of work, and we'd have to discuss the budget with Mike